cernopendata_client/downloader_class.py

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
+# This file is part of cernopendata-client.
+#
+# Copyright (C) 2020 CERN.
+#
+# cernopendata-client is free software; you can redistribute it and/or modify
+# it under the terms of the GPLv3 license; see LICENSE file for more details.
+
+"""cernopendata-client file downloading related utilities."""
+
+import sys
+
+try:
+    import requests
+
+    requests_available = True
+except ImportError:
+    requests_available = False
+
+try:
+    import pycurl
+
+    pycurl_available = True
+except ImportError:
+    pycurl_available = False
+
+
+from .printer import display_message
+
+
+class Downloader(object):
+    """Downloader class for managing pycurl and requests related utilities functions."""
```

Comment:
If we go for the object-oriented way, what about bringing it further:

- make Downloader class abstract, with common methods for showing progress etc, and with abstract method download() that will get implemented in each subclass
- create DownloaderPycurl and DownloaderRequests classes implementing download() function using each concrete downloader

This will also leave space for creating new classes:
- DownloaderXRootD class, using xrootdpyfs; this is currently enough. But later we could have another class using vanilla xrootd client (as opposed to xrootdpyfs).

Note: OT1H, the common Downloader() class could take care of both HTTP and XRootD protocols, if it makes sense to have some common functions, such as progress reporting or argument massaging. OTOH, we could also make two distinct class hierarchies, DownloaderHttp (with its DownloaderHttpPycurl, DownloaderHttpRequests subclasses) and DownloaderXrootd, which could make separation cleaner.

Reply:
The option with making `Downloader` class abstract and implementing  `DownloaderPycurl` and `DownloaderRequests`, `DownloaderXRootD` looks good to me.

I am not sure how the progress or any other methods will get implemented in `Xrootd` case so making distinct class hierarchies, DownloaderHttp (with its DownloaderHttpPycurl, DownloaderHttpRequests subclasses) and DownloaderXrootd also makes sense.

Bit confused here for choosing any of these. 😄

Reply:
After [MM](https://mattermost.web.cern.ch/it-dep/pl/y5zgyb3153gp7xmigce1fo6hne) discussion, proceeded with the second option.